src/ui/carousel/carousel.ts

```diff
@@ -0,0 +1,548 @@
+    const clampedIndex = this.getClampedIndex(index);
+    this.goToSlide(this.getSlideByIndex(clampedIndex), drivenBySync);
+  }
+
+  /**
+   * Register a function to be called when the given event is fired.
+   */
+  on(event: CarouselEvent, callback: (data: CarouselEventData) => void) {
+    this.eventManager.on(event, callback);
+  }
+
+  /**
+   * Remove a function that is called when the given event is fired.
+   */
+  off(event: CarouselEvent, callback: (data: CarouselEventData) => void) {
+    this.eventManager.off(event, callback);
+  }
+
+  /**
+   * Return the slide element at the given index in the list of slides.
+   * @param index
+   */
+  getSlideByIndex(index: number): HTMLElement {
+    return this.slides[index];
+  }
+
+  /**
+   * Return the number of slides.
+   */
+  getSlideCount(): number {
+    return this.slides.length;
+  }
+
+  /**
+   * Dispose of the carousel.
+   */
+  dispose() {
+    this.raf.dispose();
+    this.transition.dispose();
+  }
+
+  /**
+   * Clear the transition target, stop transitioning.
+   */
+  stopTransition(): void {
+    this.transitionTarget = null;
+  }
+
+  /**
+   * Clamp a given index to land within the possible slide indices.
+   * @param index
+   */
+  getClampedIndex(index: number): number {
+    const slidesLength = this.getSlides().length;
+    if (this.loop) {
+      return mathf.wrap(index, 0, slidesLength);
+    } else {
+      return mathf.clamp(0, slidesLength - 1, index);
+    }
+  }
+
+  /**
+   * Synchronize this carousel with the given carousels.
+   */
+  sync(...carousels: Carousel[]): void {
+    const masterSet = setf.merge(
+        this.syncedCarousels,
+        ...carousels.map((c) => c.syncedCarousels));
+    this.syncedCarousels = masterSet;
+    carousels.forEach((c) => c.syncedCarousels = masterSet);
+  }
+
+  /**
+   * Transition to the slide `value` slides away.
+   * @param value
+   */
+  private transitionSlidesBy(value: number): void {
+    const nextIndex =
+        this.getSlides()
+            .indexOf(this.transitionTarget || this.getActiveSlide()) +
+        value;
+    this.goToIndex(nextIndex);
+  }
+
+  /**
+   * Sync this carousel to the given index in the given carousel.
+   * @param index
+   * @param carousel
+   * @private
+   */
+  private syncTo(index: number, carousel: Carousel) {
+    // Allow for syncing with duplicate slides.
+    // If Carousel A needed its slides duplicated to meet design motion
+    // requirements, but Carousel B didn't, A's slide 2X is equivalent to B's
+    // slide X, so if B transitions to X and A is already at 2X, we can safely
+    // ignore that sync, to prevent what will seem to the user like an
+    // unnecessary transition.
+    if (
+        this.getIndex(this.transitionTarget) % carousel.getSlideCount() !==
+        index
+    ) {
+      // If the current carousel has more slides than the syncing carousel,
+      // then we want to grab the closest equivalent slide.
```

Comment:
I had a bit of trouble following the logic in this particular block of code, I wonder if the code can be simplified somehow or if you could update the comment above to clarify with an example of what an "equivalent slide" means.

Reply:
Added more comments to clarify the motivation behind this code.